src/coreComponents/physicsSolvers/wavePropagation/ElasticWaveEquationSEMKernel.hpp

```diff
@@ -0,0 +1,719 @@
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file ElasticWaveEquationSEMKernel.hpp
+ */
+
+#ifndef GEOSX_PHYSICSSOLVERS_WAVEPROPAGATION_ELASTICWAVEEQUATIONSEMKERNEL_HPP_
+#define GEOSX_PHYSICSSOLVERS_WAVEPROPAGATION_ELASTICWAVEEQUATIONSEMKERNEL_HPP_
+
+#include "finiteElement/kernelInterface/KernelBase.hpp"
+
+
+namespace geosx
+{
+
+/// Namespace to contain the elastic wave kernels.
+namespace elasticWaveEquationSEMKernels
+{
+
+struct PrecomputeSourceAndReceiverKernel
+{
+  /**
+   * @brief Check if the source point is inside an element or not
+   * @param numFacesPerElem number of face on an element
+   * @param elemCenter array containing the center of the elements
+   * @param faceNormal array containing the normal of all faces
+   * @param faceCenter array containing the center of all faces
+   * @param elemsToFaces map to get the global faces from element index and local face index
+   * @param coords coordinate of the point
+   * @return true if coords is inside the element
+   */
+
+  GEOSX_HOST_DEVICE
+  static bool
+  locateSourceElement( real64 const numFacesPerElem,
+                       real64 const (&elemCenter)[3],
+                       arrayView2d< real64 const > const faceNormal,
+                       arrayView2d< real64 const > const faceCenter,
+                       arraySlice1d< localIndex const > const elemsToFaces,
+                       real64 const (&coords)[3] )
+  {
+    //Loop over the element faces
+    localIndex prevSign = 1;
+    real64 tmpVector[3]{};
+    for( localIndex kfe = 0; kfe < numFacesPerElem; ++kfe )
+    {
+
+      localIndex const iface = elemsToFaces[kfe];
+      real64 faceCenterOnFace[3] = {faceCenter[iface][0],
+                                    faceCenter[iface][1],
+                                    faceCenter[iface][2]};
+      real64 faceNormalOnFace[3] = {faceNormal[iface][0],
+                                    faceNormal[iface][1],
+                                    faceNormal[iface][2]};
+
+      //Test to make sure if the normal is outwardly directed
+      LvArray::tensorOps::copy< 3 >( tmpVector, faceCenterOnFace );
+      LvArray::tensorOps::subtract< 3 >( tmpVector, elemCenter );
+      if( LvArray::tensorOps::AiBi< 3 >( tmpVector, faceNormalOnFace ) < 0.0 )
+      {
+        LvArray::tensorOps::scale< 3 >( faceNormalOnFace, -1 );
+      }
+
+      // compute the vector face center to query point
+      LvArray::tensorOps::subtract< 3 >( faceCenterOnFace, coords );
+      localIndex const s = computationalGeometry::sign( LvArray::tensorOps::AiBi< 3 >( faceNormalOnFace, faceCenterOnFace ));
+
+      // all dot products should be non-negative (we enforce outward normals)
+      if( prevSign * s < 0 )
+      {
+        return false;
+      }
+      prevSign = s;
+
+    }
+    return true;
+  }
```

Comment:
```suggestion
      // all dot products should be non-negative (we enforce outward normals)
      if( s < 0 )
      {
        return false;
      }

    }
    return true;
  }
```
This part is probably copied from the function `computationalGeometry::isPointInsidePolyhedron` which is used in the acoustic solver. I don't see why `prevSign` is needed. It is causing a severe bug in some cases: when the point (source or receiver) is located on an element face, some elements on the mesh can still detect it as local even if those elements are very far from the true point location. I was able to produce a such bug, which naturally leads to an incorrect seismogram.
This issue may belong to a different PR but I just wanted to point it out here because I was testing this elastic solver and ran into the problem.

Reply:
I agree prevSign is an issue. 
In fact I copied the `computationalGeometry::isPointInsidePolyhedron` for future compatibility with High Order stuff (if I do not it implies changes in a high level of GEOSX with touch every solvers and broke integrated tests).
If we do an other PR, I think we will completely change the way to locate the pouint to make it more robust, I need to speak with Stefano about that.